File: lua/test/rpm_entity_test.lua
-- Rpm entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("cloudsmith_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("RpmEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Rpm(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = rpm_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "list", "update", "load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "rpm." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set CLOUDSMITH_TEST_RPM_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local rpm_ref01_ent = client:Rpm(nil)
    local rpm_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.rpm"), "rpm_ref01"))
    rpm_ref01_data["identifier"] = setup.idmap["identifier01"]
    rpm_ref01_data["owner"] = setup.idmap["owner01"]

    local rpm_ref01_data_result, err = rpm_ref01_ent:create(rpm_ref01_data, nil)
    assert.is_nil(err)
    rpm_ref01_data = helpers.to_map(rpm_ref01_data_result)
    assert.is_not_nil(rpm_ref01_data)

    -- LIST
    local rpm_ref01_match = {
      ["identifier"] = setup.idmap["identifier01"],
      ["owner"] = setup.idmap["owner01"],
    }

    local rpm_ref01_list_result, err = rpm_ref01_ent:list(rpm_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(rpm_ref01_list_result)

    local found_item = vs.select(
      runner.entity_list_to_data(rpm_ref01_list_result),
      { id = rpm_ref01_data["id"] })
    assert.is_false(vs.isempty(found_item))

    -- UPDATE
    local rpm_ref01_data_up0_up = {
      ["identifier"] = setup.idmap["identifier"],
      ["owner"] = setup.idmap["owner"],
    }

    local rpm_ref01_markdef_up0_name = "auth_mode"
    local rpm_ref01_markdef_up0_value = "Mark01-rpm_ref01_" .. tostring(setup.now)
    rpm_ref01_data_up0_up[rpm_ref01_markdef_up0_name] = rpm_ref01_markdef_up0_value

    local rpm_ref01_resdata_up0_result, err = rpm_ref01_ent:update(rpm_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local rpm_ref01_resdata_up0 = helpers.to_map(rpm_ref01_resdata_up0_result)
    assert.is_not_nil(rpm_ref01_resdata_up0)
    assert.are.equal(rpm_ref01_resdata_up0[rpm_ref01_markdef_up0_name], rpm_ref01_markdef_up0_value)

    -- LOAD
    local rpm_ref01_match_dt0 = {}
    local rpm_ref01_data_dt0_loaded, err = rpm_ref01_ent:load(rpm_ref01_match_dt0, nil)
    assert.is_nil(err)
    assert.is_not_nil(rpm_ref01_data_dt0_loaded)

  end)
end)

function rpm_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/rpm/RpmTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read rpm test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "rpm01", "rpm02", "rpm03", "package01", "package02", "package03", "repo01", "repo02", "repo03", "identifier01", "owner01" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("CLOUDSMITH_TEST_RPM_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["CLOUDSMITH_TEST_RPM_ENTID"] = idmap,
    ["CLOUDSMITH_TEST_LIVE"] = "FALSE",
    ["CLOUDSMITH_TEST_EXPLAIN"] = "FALSE",
    ["CLOUDSMITH_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["CLOUDSMITH_TEST_RPM_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end
  if idmap_resolved["identifier"] == nil then
    idmap_resolved["identifier"] = idmap_resolved["identifier01"]
  end
  if idmap_resolved["owner"] == nil then
    idmap_resolved["owner"] = idmap_resolved["owner01"]
  end

  if env["CLOUDSMITH_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["CLOUDSMITH_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["CLOUDSMITH_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["CLOUDSMITH_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: go/test/rpm_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/cloudsmith-sdk/go"
	"github.com/voxgig-sdk/cloudsmith-sdk/go/core"

	vs "github.com/voxgig-sdk/cloudsmith-sdk/go/utility/struct"
)

func TestRpmEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Rpm(nil)
		if ent == nil {
			t.Fatal("expected non-nil RpmEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"rpm": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Rpm(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.MakeConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Rpm(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := rpmBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "rpm." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set CLOUDSMITH_TEST_RPM_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		rpmRef01Ent := client.Rpm(nil)
		rpmRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "rpm"}, setup.data), "rpm_ref01"))
		rpmRef01Data["identifier"] = setup.idmap["identifier01"]
		rpmRef01Data["owner"] = setup.idmap["owner01"]

		rpmRef01DataResult, err := rpmRef01Ent.Create(rpmRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		rpmRef01Data = core.ToMapAny(rpmRef01DataResult)
		if rpmRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

		// LIST
		rpmRef01Match := map[string]any{
			"identifier": setup.idmap["identifier01"],
			"owner": setup.idmap["owner01"],
		}

		rpmRef01ListResult, err := rpmRef01Ent.List(rpmRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		rpmRef01List, rpmRef01ListOk := rpmRef01ListResult.([]any)
		if !rpmRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", rpmRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(rpmRef01List), map[string]any{"id": rpmRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		rpmRef01DataUp0Up := map[string]any{
			"identifier": setup.idmap["identifier"],
			"owner": setup.idmap["owner"],
		}

		rpmRef01MarkdefUp0Name := "auth_mode"
		rpmRef01MarkdefUp0Value := fmt.Sprintf("Mark01-rpm_ref01_%d", setup.now)
		rpmRef01DataUp0Up[rpmRef01MarkdefUp0Name] = rpmRef01MarkdefUp0Value

		rpmRef01ResdataUp0Result, err := rpmRef01Ent.Update(rpmRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		rpmRef01ResdataUp0 := core.ToMapAny(rpmRef01ResdataUp0Result)
		if rpmRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if rpmRef01ResdataUp0[rpmRef01MarkdefUp0Name] != rpmRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", rpmRef01MarkdefUp0Name, rpmRef01ResdataUp0[rpmRef01MarkdefUp0Name])
		}

		// LOAD
		rpmRef01MatchDt0 := map[string]any{}
		rpmRef01DataDt0Loaded, err := rpmRef01Ent.Load(rpmRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if rpmRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func rpmBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "rpm", "RpmTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read rpm test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse rpm test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"rpm01", "rpm02", "rpm03", "package01", "package02", "package03", "repo01", "repo02", "repo03", "identifier01", "owner01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("CLOUDSMITH_TEST_RPM_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"CLOUDSMITH_TEST_RPM_ENTID": idmap,
		"CLOUDSMITH_TEST_LIVE":      "FALSE",
		"CLOUDSMITH_TEST_EXPLAIN":   "FALSE",
		"CLOUDSMITH_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["CLOUDSMITH_TEST_RPM_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add identifier alias for update test.
	if idmapResolved["identifier"] == nil {
		idmapResolved["identifier"] = idmapResolved["identifier01"]
	}
	// Add owner alias for update test.
	if idmapResolved["owner"] == nil {
		idmapResolved["owner"] = idmapResolved["owner01"]
	}

	if env["CLOUDSMITH_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["CLOUDSMITH_APIKEY"],
			},
			extra,
		})
		client = sdk.NewCloudsmithSDK(core.ToMapAny(mergedOpts))
	}

	live := env["CLOUDSMITH_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["CLOUDSMITH_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
